### DISCUSSION TITLE
MOSAIC simulator coupling

### DISCUSSION THREAD

**Discussion by CiscoFrisco** [eclipse/mosaic#112] [Q&A] [answered]
Hello all,

I'm currently trying to extend MOSAIC for a project that I'm developing that involves BDI agents. My idea is to use the spade-bdi python module to implement BDI agents using a python script. It would basically create new BDI agents as new vehicles register MOSAIC and then use these agents as needed.

As far as I know, to integrate a new component into MOSAIC you need to develop a federate since it's how MOSAIC can communicate with the simulator (in this case the running python script). 

So far, I defined a federate for the spade component but I don't really understand what it is doing. It is starting when MOSAIC starts but besides that, it does nothing. By looking at MOSAIC's repo I noticed that for example in your SUMO Ambassador you define an executor for launching sumo. I tried to do the same with my python script within the federate I defined, but nothing happened. 

I'm having trouble developing this federate because I do not know how I can use it to communicate with my python script. And as far as the documentation goes I think there's nothing there left that is useful for me to use. I've been relying on your repo to try to understand how you developed your other ambassadors but I'm having trouble understanding it.

How can I use an ambassador to establish communication between a vehicle application and the python script that I want to run when launching my scenario? And how can I exchange information between them?

Best regards, 
Francisco Filipe

**Comment by kschrab** [comment, marked as answer]
Hi Francisco,

as soon as the simulator you want to couple is not based on Java, you will need to implement a socket-based interface to exchange messages between the ambassador (Java class coupled directly to MOSAIC), and the simulator (your python code). Currently there are three simulators coupled to MOSAIC which use such interface: The SumoAmbassador uses the TraCI interface provided directly by SUMO. This won't help you because it is a protocol offered by SUMO itself, we just wrote the client code for that to use this API. For our other couplings of OMNeT++ and ns-3 we implemented an own socket interface based on Protobuf messages. You can find that in the 
`AbstractNetworkAmbassador` class and [the code for the federate here](https://github.com/mosaic-addons/omnetpp-federate). 

Unfortunately, we currently do _not_ offer a "general purpose" socket API which would make it a lot easier to couple external simulators. All API available currently relies on Java. If you still want to continue and couple your python based simulator, I would suggest you to use gRPC or the like.

**Reply by CiscoFrisco** [reply]
What if I use a Java Framework instead of a python module? Don't I still need to implement any socket-based communication interface? Would it simplify the coupling? I'm low on time for this project so the simpler the implementation of this MOSAIC simulator coupling the better. 

Thanks for the quick reply

**Reply by kschrab** [reply]
If you could use this as a library (jar file) you can just include it and access the methods directly, then you sure do not need a socket-based interface which would simplify the coupling a lot. 

Just create a new ambassador class which extends `AbstractFederateAmbassador`:

```java
    package org.eclipse.mosaic.example;

    public class MyAmbassador extends AbstractFederateAmbassador {

        protected MyAmbassador(AmbassadorParameter ambassadorParameter) {
            super(ambassadorParameter);
        }

        @Override
        protected void processTimeAdvanceGrant(long time) throws InternalFederateException {
            // do something periodically

            // example: request to be called again in 1 second
            rti.requestAdvanceTime(time + TIME.SECOND);
        }

        @Override
        protected void processInteraction(Interaction interaction) throws InternalFederateException {
            if (interaction instanceof VehicleUpdates) {
                for (VehicleData updated: ((VehicleUpdates) interaction).getUpdated()) {
                    String name = updated.getName();
                    GeoPoint position = updated.getPosition();
                    // example: do something with the data of vehicles
                }

                // example: send a custom interaction in 10 seconds
                rti.triggerInteraction(new MyInteraction(interaction.getTime() + 10 * TIME.SECOND));
            }
        }

        @Override
        public boolean isTimeConstrained() {
            // if true, this ambassador expects interactions to be received in timestamp order
            return true;
        }

        @Override
        public boolean isTimeRegulating() {
            // if true, it is expected that this ambassador sends time stamped interactions
            // (only relevant for optimistic synchronization)
            return true;
        }
    }
```

and furthermore a new entry in the `runtime.json`:

```json
        {
            "id": "myambassador",
            "classname": "org.eclipse.mosaic.example.MyAmbassador",
            "configuration": "myambassador_config.json",
            "priority": 50,
            "host": "local",
            "port": 0,
            "deploy": false,
            "start": false,
            "subscriptions": [
                "VehicleRegistration",
                "VehicleUpdates"
            ]
        },
```

and then you add the id `myambassador` into your `scenario_config.json`: 

```json
    "federates": {
        "application": true,
        "mapping": true,
        "myambassador": true,
        "sumo": true
    }
```

**Reply by CiscoFrisco** [reply]
Thanks. That's what I did as of now, create a new ambassador, add it to the runtime.json and to the scenario_config.json. So If I import the library in this ambassador file and call the methods in it, would it be correctly implemented then?

**Reply by CiscoFrisco** [reply]
One last question: once the ambassador is implemented, how can I communicate with it from my custom applications? Is that even possible? My end goal here is to have my vehicle applications exchange information with another simulator through this ambassador as they need. Can I develop methods in the ambassador and call those methods in my application?

**Reply by kschrab** [reply]
As application simulation and any other simulators are decoupled from each other, you will need special interactions to exchange information. In order to send interactions from an application to an ambassador, you simply call `getOs().sendInteractionToRti(..)` from within your application and put you custom interaction into it. Do not forget to update the `runtime.json` so that your ambassador subscribes to this new interaction.

If you want to send an interaction from your ambassador to an application, you would need to do the following:

1. Add a new interaction which extends the abstract class `org.eclipse.mosaic.interactions.application.ApplicationInteraction`.
2. On ambassador side, you create an instance of this interaction and send it to the RTI: `rti.triggerInteraction(...);`. If you pass a specific vehicle id to this interaction, only the applications of this vehicle will receive the interaction. If you pass `null`, all vehicles will receive it.
3. Let your application implement the interface `org.eclipse.mosaic.fed.application.app.api.MosaicApplication` and it will receive this interaction with `onInteractionReceived(..)`.

**Reply by CiscoFrisco** [reply]
Thank you very much. During the weekend I ended up diving deeper through the repo to find more information and actually implemented the communication using the interactions as you said. So now I know it's well implemented. Thanks for the quick response

**Reply by kschrab** [reply]
Glad to hear! :)
